keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1201 @@
+not feasible. Also, users can write ratcheting logic directly in CEL as a 
+transition rule if that was required.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+At a high level, CR validation in kubernetes has four components:
+1. `kube-openapi`-based `Validate` using `*spec.Schema`
+2. StructuralSchema-based Meta, TypeMeta, ListSet/Map validation
+3. CEL Validators using a CEL schema interface type
+4. Embedded Objects
+
+Each of these validation paths will need to have their `ValidateUpdate(old, new)`
+codepath modified to:
+
+1. Take an option to enable the feature
+2. Ignore errors which occur both on the `old` and `new` objects
+3. Return ignored errors in a separate list
+
+### `kube-openapi` changes
+
+The `kube-openapi` schema validator validates the standard OpenAPI/JSONSchema
+elements of the CRD schema. This is basically everything except CEL rules,
+and listset/map types.
+
+The current SchemaValidator will be wrapped with a `RatchetingSchemaValidator`
+which compares old/new values and ignores all errors on its subpath if they are
+equal.
+
+Supporting validation code within kube-openapi which spawn a new `SchemaValidator`
+will need to be refactored to find the old value and use `RatchetingSchemaValidator`
+when `RatchetingSchemaValidator.ValidateUpdate` is in its callstack.
+
+anyOf, allOf, not, allOf sections of logic will remain using the normal 
+`SchemaValidator`.
+
+#### Nested Value Validations 
+
+OpenAPI-schemas support nested validations:
+- not
+- oneOf
+- anyOf
+- allOf
+
+
+To keep the semantics simple and implementation clear, none of these constructs
+nor any of their subschemas will be ratcheted. Our intution is that in the wild
+CRD authors do not make much use of them, and when they do the logic is 
+complicated and references multiple fields.
+
+### Structural-Schema-based validation changes
+
+The two structural-schema based validation methods will need to
+be refactored to support `ValidateUpdate` returning separate lists of errors.
+
+[`ValidateListSetsAndMaps(fldPath *field.Path, s *schema.Structural, obj map[string]interface{}) field.ErrorList`](https://github.com/kubernetes/kubernetes/blob/694698ca3801fe5aa258d58c1e27ac7f88c29d35/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/listtype/validation.go#L28)
+
+[`objectmeta.Validate(pth *field.Path, obj interface{}, s *structuralschema.Structural, isResourceRoot bool) field.ErrorList`](https://github.com/kubernetes/kubernetes/blob/92042fe6eac61b6dd4092f5c6737050a4a1c43e5/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/validation.go#L32)
```

Comment:
Not sure about this statement. This is a core validation function in kube. I don't think we want to touch it to add general ratcheting (it would have larger impact than what this KEP describes). On the other hand, we could validate `old` in case of error on `obj` and ratchet on the whole ObjectMeta level.

Reply:
Added a note that for alpha we will only do object level ratcheting of the metadata due to the risk of changing this core function